File: selected_coding_interview/codes/java/lc_226_invert_binary_tree/lc_226_invert_binary_tree_s2.java
/*
* File: lc_226_invert_binary_tree_s2.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_226_invert_binary_tree;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public TreeNode invertTree(TreeNode root) {
            if (root == null) return null;
            Stack<TreeNode> stack = new Stack<>() {{ add(root); }};
            while (!stack.isEmpty()) {
                TreeNode node = stack.pop();
                if (node.left != null) stack.add(node.left);
                if (node.right != null) stack.add(node.right);
                TreeNode tmp = node.left;
                node.left = node.right;
                node.right = tmp;
            }
            return root;
        }
    }

public class lc_226_invert_binary_tree_s2 {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1: Basic binary tree
        TreeNode root = TreeNode.arrToTree(new Integer[]{3, 9, 20, null, null, 15, 7});

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.invertTree(root);
        System.out.println(result);

    }
}

File: selected_coding_interview/codes/cpp/lc_226_invert_binary_tree/lc_226_invert_binary_tree_s2.cpp
/*
 * File: lc_226_invert_binary_tree_s2.cpp
 * Created Time: 2025-12-30
 * Author: krahets
 */

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if (root == nullptr) return nullptr;
        stack<TreeNode*> stack;
        stack.push(root);
        while (!stack.empty())
        {
            TreeNode* node = stack.top();
            stack.pop();
            if (node->left != nullptr) stack.push(node->left);
            if (node->right != nullptr) stack.push(node->right);
            TreeNode* tmp = node->left;
            node->left = node->right;
            node->right = tmp;
        }
        return root;
    }
};

int main() {
    // ======= Test Case =======
    TreeNode* root = vectorToTree({4, 2, 7, 1, 3, 6, 9});

    // ====== Driver Code ======
    Solution* slt = new Solution();
    TreeNode* res = slt->invertTree(root);
    PrintUtil::printTree(res);

    return 0;
}
